fio_wrapper/fio_wrapper.py

```diff
@@ -0,0 +1,163 @@
+import logging
+import hashlib
+import json
+from copy import deepcopy
+from fio_analyzer import Fio_Analyzer
+from utils.py_es_bulk import streaming_bulk
+from utils.common_logging import setup_loggers
+import trigger_fio
+
+logger = logging.getLogger("fio_wrapper")
+
+es_log = logging.getLogger("elasticsearch")
+es_log.setLevel(logging.CRITICAL)
+urllib3_log = logging.getLogger("urllib3")
+urllib3_log.setLevel(logging.CRITICAL)
+
+setup_loggers("fio_wrapper", logging.DEBUG)
+
+
+def main():
+    
+    #collect arguments
+    parser = argparse.ArgumentParser(description="fio-d Wrapper script")
+    parser.add_argument(
+        'hosts', help='Provide host file location')
+    parser.add_argument(
+        'job', help='path to job file')
+    parser.add_argument(
+        '-s', '--sample', type=int, default=1, help='number of times to run benchmark, defaults to 1')
+    parser.add_argument(
+        '-d', '--dir', help='output parent directory', default=os.path.dirname(os.getcwd()))
+    args = parser.parse_args()
+    
+    args.index_results = False
+    args.prefix = "ripsaw-fio-"
+    es={}
+    if "es" in os.environ:
+        es['server'] = os.environ["es"]
+        es['port'] = os.environ["es_port"]
+        args.prefix = os.environ["es_index"]
+        args.index_results = True
+    
+        es = elasticsearch.Elasticsearch([
+        {'host': es['server'],'port': es['port'] }],send_get_body_as='POST')
+    
+    #call py es bulk using a process generator to feed it ES documents
+    res_beg, res_end, res_suc, res_dup, res_fail, res_retry  = streaming_bulk(es, process_generator(args))
+               
+    # set up a standard format for time
+    FMT = '%Y-%m-%dT%H:%M:%SGMT'
+    start_t = time.strftime('%Y-%m-%dT%H:%M:%SGMT', time.gmtime(res_beg))
+    end_t = time.strftime('%Y-%m-%dT%H:%M:%SGMT', time.gmtime(res_end))
+    start_t = datetime.datetime.strptime(start_t, FMT)
+    end_t = datetime.datetime.strptime(end_t, FMT)
+    
+    #get time delta for indexing run
+    tdelta = end_t - start_t
+    logger.info("Duration of indexing - %s" % tdelta)
+    logger.info("Indexed results - %s success, %s duplicates, %s failures, with %s retries." % (res_suc,
+                                                                                                res_dup,
+                                                                                                res_fail,
+                                                                                                res_retry)) 
+
+def process_generator(args):
+     
+    object_generator = process_data(args)
+    
+    for object in object_generator:
+        for action, index in object.emit_actions():
+            
+            es_valid_document = { "_index": index,
+                                  "_type": "result",
+                                  "_op_type": "create",
+                                  "_source": action,
+                                  "_id": "" }
```

Comment:
why are we generating id here instead of having it generated directly on serverside?

Reply:
If one does not generate their own id + use the create method elasticsearch will perform this action, thus causing an increase in computes to index on the ES server/cluster. using the create op_type and creating the id while indexing allows for faster indexing and eliminates the chance of duplication. regardless of likelihood of duplication, this approach safeguards against if there is ever a re-indexing of an archive.